### PR TITLE
Status Improvements

### LINKED ISSUE #1571: SetStatus must store the StatusCode string value.
Instead of 0,1,2, it should store Unset,Ok,Error.
https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/src/OpenTelemetry.Api#setting-status

### PR DESCRIPTION
**Pull request by CodeBlanch**
Fixes #1571.

## Change: Fix StatusCode being sent as integer

I noticed this helping someone out on Gitter:

image

* We're sending StatusCode (otel.status_code) as the enum integer value.
* We're sending the Description (otel.status_description) on every http span (set to the reason phrase if it is available).

The [Status spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status) defines the string names and doesn't mention the numeric values.

New version looks like this:

image

* We send the string representation of the StatusCode (otel.status_code).
* We no longer send the Description (otel.status_description) on every http span. My reasoning for that:
   1) If http reason phrase was part of the spec, it would be on http semantic conventions.
   2) [.NET Core HttpResonse doesn't even have StatusDescription, because it isn't part of HTTP/2](https://github.com/aspnet/HttpAbstractions/issues/395).

Note: We will continue to send a Description (otel.status_description) for http spans if an exception is thrown.

## Change: JaegerExporter will now set error flag

Jaeger spec actually has a way to indicate that a span errored out. With the previous `Status` spec, it was really hard to know what was an error, so we didn't set the flag. But now we have a dedicated `Error` status so we can turn this on:

image

## Change: ZipkinExporter will now set error flag

Zipkin spec actually has a way to indicate that a span errored out. With the previous `Status` spec, it was really hard to know what was an error, so we didn't set the flag. But now we have a dedicated `Error` status so we can turn this on:

image

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes

### CODE REVIEW

**Review by reyang — APPROVED** [review]
LGTM.

**Comment by CodeBlanch** [comment]
For anyone curious, I did benchmark switch statement vs dictionary lookup for the conversions to/from StatusCode/string. The lower you are in the switch, the slower the perf. But with the 3/4 cases, it's always faster than the dictionary.

|                 Method | StatusCode |      Mean |     Error |    StdDev |    Median |
|----------------------- |----------- |----------:|----------:|----------:|----------:|
|       **DictionaryLookup** |      **Error** | **17.956 ns** | **0.3061 ns** | **0.3525 ns** | **17.789 ns** |
|           SwitchLookup |      Error |  5.803 ns | 0.0816 ns | 0.0724 ns |  5.840 ns |
|       **DictionaryLookup** |         **Ok** | **15.594 ns** | **0.1286 ns** | **0.1074 ns** | **15.593 ns** |
|           SwitchLookup |         Ok |  3.236 ns | 0.0424 ns | 0.0376 ns |  3.235 ns |
|       **DictionaryLookup** |    **Unknown** | **13.100 ns** | **0.1269 ns** | **0.1125 ns** | **13.101 ns** |
|           SwitchLookup |    Unknown |  4.957 ns | 0.1240 ns | 0.1613 ns |  4.890 ns |
|       **DictionaryLookup** |      **Unset** | **18.120 ns** | **0.3561 ns** | **0.3331 ns** | **18.072 ns** |
|           SwitchLookup |      Unset |  7.551 ns | 0.0665 ns | 0.0556 ns |  7.559 ns |

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ public static void SetStatus(this Activity activity, Status status)
         {
             Debug.Assert(activity != null, "Activity should not be null");
 
-            activity.SetTag(SpanAttributeConstants.StatusCodeKey, (int)status.StatusCode);
+            activity.SetTag(SpanAttributeConstants.StatusCodeKey, StatusHelper.GetStringNameForStatusCode(status.StatusCode));
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:43]: can you add the type change in changelog.md for API?

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:43]: Updated

**Comment by reyang** [comment]
@CodeBlanch I guess the string comparison call might be the biggest contributor to the switch perf.

```csharp
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int? GetStatusCodeForStringName(string statusCodeName)
        {
            return statusCodeName switch
            {
                "Unset" => 0,
                "Error" => 2,
                "Ok" => 1,
                _ => null,
            };
        }
```

would translate to:

```
	IL_0000: ldarg.0
	IL_0001: brfalse.s IL_004a

	IL_0003: ldarg.0
	IL_0004: ldstr "Unset"
	IL_0009: call bool [System.Runtime]System.String::op_Equality(string, string)
	IL_000e: brtrue.s IL_002c

	IL_0010: ldarg.0
	IL_0011: ldstr "Error"
	IL_0016: call bool [System.Runtime]System.String::op_Equality(string, string)
	IL_001b: brtrue.s IL_0036

	IL_001d: ldarg.0
	IL_001e: ldstr "Ok"
	IL_0023: call bool [System.Runtime]System.String::op_Equality(string, string)
	IL_0028: brtrue.s IL_0040
```

**Review by cijothomas — APPROVED** [review]

**Comment by reyang** [comment]
So if the goal is to get the maximum performance, we can potentially look at the following steps:
1. avoid map/string hash at all.
2. avoid string comparison API, given the string is very short - and we could also avoid going through the locale (e.g. Ordinal)

**Comment by CodeBlanch** [comment]
Thanks @reyang I'll see what I can do to squeeze a bit more perf out of it.

**Pull request merged**